README.md
# **MarkItAsDone Landings Install and Develop Manual**


## **Quick Lunch**
If you want have a quick start, you can start with the `dist` folder. just go to the `dist` folder and apply your changes to it. To apply changes to the content, simply open the `index.html` file and replace your text with the text. Also, to change the images, put your pictures in the image folder according to the size specified.

In the `dist` directory, in addition to the `images` folder, there are three other folders:
```
- fonts 
- styles 
- scripts
```
These folders include font, style, and script for landing pages. If you need to change them, just read the installation and use of landing steps.

## **Prerequisites**
You may need to install a few assets before you can get started, such as Node.

### [Node.js](https://nodejs.org)

Bring up a terminal and type `node --version`.
Node should respond with a version at or above 0.10.x.

## **Install**

Next, install the local dependencies MarkItAsDone Landing requires:

```sh
$ npm install
```

Note: if you have the [Yarn](https://yarnpkg.com/) package manager installed, you can just run `yarn`.
Landing Page includes a yarn.lock file that will be used here.

That's it! You should now have everything needed to use the MarkItAsDone Landing.

---

**You may also want to get used to some of the commands available.**

## Watch For Changes & Automatically Refresh Across Devices

```sh
$ npm start
# or
$ yarn start
```

This outputs an IP address you can use to locally test and another that can be used on devices
connected to your network.

## Build & Optimize

```sh
$ npm run build
# or
$ yarn build
```

Build and optimize the current project, ready for deployment.
This includes linting as well as image, script, stylesheet and HTML optimization and minification.


## Serve the Fully Built & Optimized Site

```sh
$ npm run demo
# or
$ yarn demo
```

This outputs an IP address you can use to locally test and another that can be used on devices
connected to your network.

---

## .babelrc

[.babelrc](https://babeljs.io/docs/usage/babelrc/) is a configuration file for passing options to [Babel](https://babeljs.io) - the ES2015 transpiler recommended for writing next-generation JavaScript in Web Starter Kit. 

## .editorconfig

[EditorConfig](http://editorconfig.org/) is a file format and collection of text editor plugins for maintaining consistent coding styles between different editors and IDEs.

## gulpfile.babel.js

[Gulp](http://gulpjs.com) is a streaming build system that allows you to automate tedious development tasks. Compared with other build systems, such as Grunt, gulp uses Node.js streams as a means to automate tasks, thereby removing the need to create intermediate files when transforming source files. 

In gulp, you would install plugins, that do one thing and do it well, and construct a 'pipeline' by connecting them to each other. A `gulpfile` allows you to put together tasks that use plugins to accomplish a task like minification. 

`gulpfile.babel.js` is a gulpfile written in ES2015. The `babel` portion of the name refers to its use of the [Babel](https://babeljs.io) transpiler for enabling ES2015 code to run there.

## src/scripts/main.js

This is a file where your custom JavaScript can go. 

## src/styles/main.scss

This is a file where your custom SCSS can go. You can place any Sass you wish to have compiled into the `styles` directory.

## src/manifest.json

`manifest.json` contains a [Web Application Manifest](https://w3c.github.io/manifest/) - a simple JSON file that gives you the ability to control how your app appears to the user in the areas that they would expect to see apps (for example the mobile home screen). In here you can control what the user can launch and more importantly how they can launch it. 

For more information on the manifest, see [Web Fundamentals](https://developers.google.com/web/updates/2014/11/Support-for-installable-web-apps-with-webapp-manifest-in-chrome-38-for-Android).

## src/manifest.webapp

`manifest.webapp` refers to the proprietary [Firefox OS manifest format](https://developer.mozilla.org/en-US/Apps/Build/Manifest), and not the W3C [manifest spec](https://w3c.github.io/manifest/), designed for cross-browser open web applications. 

The Firefox OS app manifest provides information about an app (such as name, author, icon, and description) and a list of Web APIs that your app needs.

This manifest included in Web Starter Kit until Firefox OS switches to using the manifest spec instead.

## package.json

A [package.json](https://docs.npmjs.com/files/package.json) file is used to specify project tooling dependencies from [npm](http://npmjs.org) - the Node package manager. When you run `npm install`, `package.json` is read to discover what packages need to be installed. 

`package.json` can also contain other metadata such as a project description, version, license and configuration information.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2019-01-01
### Added
- Initial Release

gulpfile.babel.prod.js
'use strict';

// This gulpfile makes use of new JavaScript features.
// Babel handles this without us having to do anything. It just works.
// You can read more about the new JavaScript features here:
// https://babeljs.io/docs/learn-es2015/

import gulp from 'gulp';
import del from 'del';
import runSequence from 'run-sequence';
import browserSync from 'browser-sync';
import gulpLoadPlugins from 'gulp-load-plugins';

const $ = gulpLoadPlugins();
const reload = browserSync.reload;

// Lint JavaScript
gulp.task('lint', () =>
  gulp.src(['src/scripts/**/*.js', '!node_modules/**'])
    .pipe($.eslint())
    .pipe($.eslint.format())
    .pipe($.if(!browserSync.active, $.eslint.failAfterError()))
);

// Optimize images
gulp.task('images', () =>
  gulp.src('src/images/**/*')
    .pipe($.cache($.imagemin({
      progressive: true,
      interlaced: true
    })))
    .pipe(gulp.dest('dist/images'))
    .pipe($.size({title: 'images'}))
);

// Copy all files at the root level (src)
gulp.task('copy', () =>
  gulp.src([
    'src/*',
    '!src/*.html',
    'node_modules/apache-server-configs/dist/.htaccess'
  ], {
    dot: true
  }).pipe(gulp.dest('dist'))
    .pipe($.size({title: 'copy'}))
);

// Copy all files at the root level (src)
gulp.task('fonts', () =>
  gulp.src(
    'src/fonts/**/*').pipe(gulp.dest('dist/fonts/'))
    .pipe($.size({title: 'copy fonts'}))
);

// Compile and automatically prefix stylesheets
gulp.task('styles', () => {
  const AUTOPREFIXER_BROWSERS = [
    'ie >= 10',
    'ie_mob >= 10',
    'ff >= 30',
    'chrome >= 34',
    'safari >= 7',
    'opera >= 23',
    'ios >= 7',
    'android >= 4.4',
    'bb >= 10'
  ];

  // For best performance, don't add Sass partials to `gulp.src`
  return gulp.src([
    'src/styles/**/*.scss',
    'src/styles/**/*.css'
  ])
    .pipe($.newer('.tmp/styles'))
    .pipe($.sass({
      outputStyle: 'expanded',
      precision: 10,
      includePaths: ['.']
    }).on('error', $.sass.logError))
    .pipe($.autoprefixer(AUTOPREFIXER_BROWSERS))
    .pipe($.size({title: 'styles'}))
    .pipe(gulp.dest('.tmp/styles'));
});

// Concatenate and minify JavaScript. Optionally transpiles ES2015 code to ES5.
// to enable ES2015 support remove the line `"only": "gulpfile.babel.js",` in the
// `.babelrc` file.
gulp.task('scripts', () =>
  gulp.src([
    // Note: Since we are not using useref in the scripts build pipeline,
    //       you need to explicitly list your scripts here in the right order
    //       to be correctly concatenated
    './src/scripts/main.js'
    // Other scripts
  ])
    .pipe($.newer('.tmp/scripts'))
    .pipe($.sourcemaps.init())
    .pipe($.babel())
    .pipe($.sourcemaps.write())
    .pipe(gulp.dest('.tmp/scripts'))
    .pipe($.concat('main.min.js'))
    .pipe($.uglify({preserveComments: 'some'}))
    // Output files
    .pipe($.size({title: 'scripts'}))
    .pipe(gulp.dest('dist/scripts'))
    .pipe(gulp.dest('.tmp/scripts'))
);

// Scan your HTML for assets & optimize them
gulp.task('html', () => {
  return gulp.src('src/**/*.html')
    .pipe($.useref({
      searchPath: ['.', '.tmp', 'src']
    }))

    .pipe($.if('*.css', $.cssnano({safe: true, autoprefixer: false})))
    .pipe($.if('*.js', $.uglify({preserveComments: 'some'})))

    // Minify any HTML
    .pipe($.if('*.html', $.htmlmin({
      removeComments: true,
      collapseWhitespace: true,
      collapseBooleanAttributes: true,
      removeAttributeQuotes: true,
      removeRedundantAttributes: true,
      removeEmptyAttributes: true,
      removeScriptTypeAttributes: true,
      removeStyleLinkTypeAttributes: true,
      removeOptionalTags: true
    })))
    // Output files
    .pipe($.if('*.html', $.size({title: 'html', showFiles: true})))
    .pipe(gulp.dest('dist'));
});

// Clean output directory
gulp.task('clean', () => del(['.tmp', 'dist/*', '!dist/.git'], {dot: true}));

// Watch files for changes & reload
gulp.task('serve', ['scripts', 'styles'], () => {
  browserSync({
    notify: false,
    // Customize the Browsersync console logging prefix
    logPrefix: 'WSK',
    // Allow scroll syncing across breakpoints
    scrollElementMapping: ['main', '.mdl-layout'],
    // Run as an https by uncommenting 'https: true'
    // Note: this uses an unsigned certificate which on first access
    //       will present a certificate warning in the browser.
    // https: true,

    server: {
      baseDir: ['.tmp', 'src'],
      routes: {
        '/node_modules': 'node_modules'
      }
    },
    port: 3000
  });

  gulp.watch(['src/**/*.html'], reload);
  gulp.watch(['src/styles/**/*.{scss,css}'], ['styles', reload]);
  gulp.watch(['src/scripts/**/*.js'], ['lint', 'scripts', reload]);
  gulp.watch(['src/images/**/*'], reload);
});

// Build and serve the output from the dist build
gulp.task('serve:dist', ['default'], () =>
  browserSync({
    notify: false,
    logPrefix: 'WSK',
    // Allow scroll syncing across breakpoints
    scrollElementMapping: ['main'],
    // Run as an https by uncommenting 'https: true'
    // Note: this uses an unsigned certificate which on first access
    //       will present a certificate warning in the browser.
    // https: true,
    server: 'dist',
    port: 3001
  })
);

// Build production files, the default task
gulp.task('default', ['clean'], cb =>
  runSequence(
    'styles',
    ['lint', 'html', 'scripts', 'images', 'copy', 'fonts'],
    cb
  )
);

// Load custom tasks from the `tasks` directory
// Run: `npm install --save-dev require-dir` from the command-line
// try { require('require-dir')('tasks'); } catch (err) { console.error(err); }


gulpfile.babel.js
'use strict';

// This gulpfile makes use of new JavaScript features.
// Babel handles this without us having to do anything. It just works.
// You can read more about the new JavaScript features here:
// https://babeljs.io/docs/learn-es2015/

const { src, dest, parallel, series, watch } = require('gulp');
const del = require('del');
const babel = require('gulp-babel');
const browserSync = require('browser-sync').create();
var fs = require('fs');
var pkg = JSON.parse(fs.readFileSync('./package.json'));
const imagemin = require('gulp-imagemin');
const argv = require('yargs').argv;
const $ = require('gulp-load-plugins')();
const reload = browserSync.reload;
const isDemo = argv.build === 'demo';
const terser = require('gulp-terser');
const zipFileName = `markitasdone.com__${pkg.name}__v${pkg.version}`;

// Lint JavaScript
function lint() {
  return src(['src/scripts/**/*.js', '!node_modules/**'])
    .pipe($.eslint())
    .pipe($.eslint.format())
    .pipe($.if(!browserSync.active, $.eslint.failAfterError()));
}
exports.lint = lint;

// Optimize images
function images() {
  return src('src/images/**/*')
    .pipe(imagemin([
      imagemin.gifsicle({ interlaced: true }),
      imagemin.mozjpeg({ progressive: true }),
      imagemin.optipng({ optimizationLevel: 5 }),
      imagemin.svgo({
        plugins: [
          { removeViewBox: true },
          { cleanupIDs: false }
        ]
      })
    ]))
    .pipe(dest('dist/images'))
    .pipe($.size({ title: 'images' }));
}
exports.images = images;

// Copy all files at the root level (src)
function copy() {
  return src(
      [
        'src/*',
        '!src/*.html',
        '!src/images-placeholder'
      ],
      {
        dot: true
      }
    )
    .pipe(dest('dist'))
    .pipe($.size({ title: 'copy' }));
}
exports.copy = copy;

// Copy all files at the root level (src)
function fonts(){
  return src('src/fonts/**/*')
    .pipe(dest('dist/fonts/'))
    .pipe($.size({ title: 'copy fonts' }))
}
exports.fonts = fonts;

// Copy all files at the root level (src)
function placeholder(){
  return src('src/images-placeholder/**/*')
    .pipe(dest('dist/images/'))
    .pipe($.size({ title: 'copy fonts' }));
}

exports.placeholder = placeholder;

// Compile and automatically prefix stylesheets
function styles() {
  const AUTOPREFIXER_BROWSERS = [
    'ie >= 10',
    'ie_mob >= 10',
    'ff >= 30',
    'chrome >= 34',
    'safari >= 7',
    'opera >= 23',
    'ios >= 7',
    'android >= 4.4',
    'bb >= 10'
  ];

  // For best performance, don't add Sass partials to `gulp.src`
  return src(['src/styles/**/*.scss', 'src/styles/**/*.css'])
    .pipe($.newer('.tmp/styles'))
    .pipe(
      $.sass({
        outputStyle: 'expanded',
        precision: 10,
        includePaths: ['.']
      }).on('error', $.sass.logError)
    )
    .pipe($.autoprefixer(AUTOPREFIXER_BROWSERS))
    .pipe($.size({ title: 'styles' }))
    .pipe(dest('.tmp/styles'));
}

exports.styles = styles;

// Concatenate and minify JavaScript. Optionally transpiles ES2015 code to ES5.
// to enable ES2015 support remove the line `'only': 'gulpfile.babel.js',` in the
// `.babelrc` file.
function scripts() {
  return src([
      // Note: Since we are not using useref in the scripts build pipeline,
      //       you need to explicitly list your scripts here in the right order
      //       to be correctly concatenated
      './src/scripts/main.js'
      // Other scripts
    ])
    .pipe($.newer('.tmp/scripts'))
    .pipe($.sourcemaps.init())
    .pipe(babel({
      presets: ['@babel/env']
    }))
    .pipe($.sourcemaps.write())
    .pipe(dest('.tmp/scripts'))
    .pipe($.concat('main.min.js'))
    .pipe(terser())
    // Output files
    .pipe($.size({ title: 'scripts' }))
    .pipe(dest('dist/scripts'))
    .pipe(dest('.tmp/scripts'));
}

exports.scripts = scripts;

// Scan your HTML for assets & optimize them
function html() {
  return ( src('src/**/*.html')
      .pipe(
        $.useref({
          searchPath: ['.', '.tmp', 'src']
        })
      )

      .pipe($.if('*.css', $.cssnano({ safe: true, autoprefixer: false })))
      .pipe($.if('*.js', terser()))

      // Minify any HTML
      .pipe(
        $.if(
          file => $.match(file, '*.html') && isDemo,
          $.htmlmin({
            removeComments: true,
            collapseWhitespace: true,
            collapseBooleanAttributes: true,
            removeAttributeQuotes: true,
            removeRedundantAttributes: true,
            removeEmptyAttributes: true,
            removeScriptTypeAttributes: true,
            removeStyleLinkTypeAttributes: true,
            removeOptionalTags: true
          })
        )
      )
      // Output files
      .pipe($.if('*.html', $.size({ title: 'html', showFiles: true })))
      .pipe(dest('dist'))
  );
}

exports.html = html;

// Clean output directory
function clean(cb) {
  del(['.tmp', 'dist/*', '!dist/.git'], { dot: true });
  cb();
}

exports.clean = clean;

// Watch files for changes & reload

// const serve = series(parallel(styles, scripts), function () {
//   browserSync.init({
//     notify: false,
//     // Customize the Browsersync console logging prefix
//     logPrefix: 'WSK',
//     // Allow scroll syncing across breakpoints
//     scrollElementMapping: ['main', '.mdl-layout'],
//     // Run as an https by uncommenting 'https: true'
//     // Note: this uses an unsigned certificate which on first access
//     //       will present a certificate warning in the browser.
//     // https: true,

//     server: {
//       baseDir: ['.tmp', 'src'],
//       routes: {
//         '/node_modules': 'node_modules'
//       }
//     },
//     port: 3000
//   });

//   watch('src/**/*.html').on('change', reload);
//   watch('src/styles/**/*.{scss,css}', styles).on('change', reload);
//   watch('src/scripts/**/*.js', parallel(lint, scripts)).on('change', reload);
//   watch('src/images/**/*').on('change', reload);
// });

function start() {
  browserSync.init({
    notify: false,
    // Customize the Browsersync console logging prefix
    logPrefix: 'WSK',
    // Allow scroll syncing across breakpoints
    scrollElementMapping: ['main', '.mdl-layout'],
    // Run as an https by uncommenting 'https: true'
    // Note: this uses an unsigned certificate which on first access
    //       will present a certificate warning in the browser.
    // https: true,

    server: {
      baseDir: ['.tmp', 'src'],
      routes: {
        '/node_modules': 'node_modules'
      }
    },
    port: 3000
  });

  watch('src/**/*.html').on('change', reload);
  watch('src/styles/**/*.{scss,css}', styles).on('change', reload);
  watch('src/scripts/**/*.js', parallel(lint, scripts)).on('change', reload);
  watch('src/images/**/*').on('change', reload);
}
exports.start = start;

exports.serve = series(parallel(styles, scripts), start);

// Build and serve the output from the dist build
function servedist() {
  browserSync.init({
    notify: false,
    logPrefix: 'WSK',
    // Allow scroll syncing across breakpoints
    scrollElementMapping: ['main'],
    // Run as an https by uncommenting 'https: true'
    // Note: this uses an unsigned certificate which on first access
    //       will present a certificate warning in the browser.
    // https: true,
    server: 'dist',
    port: 3001
  })
}

exports.servedist = servedist;


// Zip demo files
function zip() {
  return src('dist/**/*', { dot: true })
    .pipe($.zip(`${zipFileName}${isDemo ? '__DEMO' : ''}.zip`))
    .pipe(dest('.tmp'));
}

exports.zip = zip;


// Zip Release files
function ziprelease() {
  return src('.tmp/release/**/*', { dot: true })
    .pipe($.zip(`${zipFileName}.zip`))
    .pipe(dest('.tmp'));
}

exports.ziprelease = ziprelease;


// Build production files, the default task
exports.default = series(clean, styles, lint, html, scripts, images, copy, fonts, zip);

/**
 * Generrate release commands
 * @return {Array} commands
 */
function getReleaseCommands() {
  const basePath = './.tmp/release';

  const cp = [
    'dist',
    'src',
    '.babelrc',
    '.gitattributes',
    '.gitignore',
    'gulpfile.babel.prod.js',
    'package.prod.json',
    'README.md',
    'CHANGELOG.md',
    'LICENSE'
  ].map(p => `cp -rf ./${p} ${basePath}/${p.replace('.prod', '')}`);

  return [
    `mkdir ${basePath}`,
    ...cp,
    `rm -rf ${basePath}/src/images`,
    `mv ${basePath}/src/images-placeholder ${basePath}/src/images`
  ];
}

exports.getReleaseCommands = getReleaseCommands;

// Create release directory into .tmp for generating release zip
// gulp.task('cp-release', run(getReleaseCommands()));

// // Build release zip
// gulp.task('release', ['clean'], cb => {
//   runSequence(
//     'styles',
//     ['lint', 'html', 'scripts', 'cp-placeholders', 'copy', 'fonts'],
//     'cp-release',
//     'zip-release',
//     cb
//   );
// });
